### PR TITLE
add support for reduction = None and "sum" in Cross Entropy

### PR DESCRIPTION
**Pull request by shivam15s**
## Summary

Same as title #83 

## Testing Done

- Hardware Type: <BLANK>
- [ ] run `make test` to ensure correctness
- [ ] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by ByronHsu** [comment]
merge after 0.1.2 release

File: src/liger_kernel/ops/cross_entropy.py [hunk 1]
```diff
@@ -14,6 +14,7 @@ def liger_cross_entropy_kernel(
     n_cols,
     n_non_ignore,
     ignore_index,
+    reduction: tl.constexpr,
```
Comment (ByronHsu) [inline src/liger_kernel/ops/cross_entropy.py:17]:
```suggestion
    reduction: tl.constexpr, # set it as constexpr since reduction is always known at compile time
```

**Review by ByronHsu — COMMENTED** [review]
overall lg. will do another pass after the release

**Comment by lancerts** [comment]
can we resolve the conflicts @shivam15s

**Comment by ByronHsu** [comment]
please also solve https://github.com/linkedin/Liger-Kernel/issues/177

File: src/liger_kernel/ops/cross_entropy.py [hunk 2]
```diff
@@ -82,7 +84,11 @@ def liger_cross_entropy_kernel(
         X_block = tl.load(
             X_ptr + X_offsets, mask=X_offsets < n_cols, other=float("-inf")
         )
-        X_block = (tl.exp(X_block - m) / d) / (n_non_ignore)
+        if reduction == "mean":
```
Comment (ByronHsu) [inline src/liger_kernel/ops/cross_entropy.py:87]: we have to change the formula at 79 and 80 too

File: src/liger_kernel/ops/fused_linear_cross_entropy.py [hunk 3]
```diff
@@ -109,7 +119,9 @@ def forward(ctx, _input, weight, target, bias=None, ignore_index=-100):
             # thus grad_input[start_idx: end_idx] should be of shape: chunk_size x H
             # additionally, since we are chunking the inputs, observe that the loss and gradients are calculated only
             # on `n_non_ignore` tokens. However, the gradient of the input should be calculated for all tokens.
-            # Thus, we need an additional scaling factor of (n_non_ignore/total_n_non_ignore) to scale the gradients.
+            # Thus, we need an additional scaling factor of (n_non_ignore/total_n_non_ignore) to scale the loss and gradients.
+
+            loss_1d[start_idx:end_idx] = loss_1d_slice * n_non_ignore
```
Comment (ByronHsu) [inline src/liger_kernel/ops/fused_linear_cross_entropy.py:124]: for "mean", i think we need to scale by `n_non_ignore / total_n_non_ignore`?
for others, do we need to scale? no?

File: src/liger_kernel/ops/fused_linear_cross_entropy.py [hunk 4]
```diff
@@ -109,7 +119,9 @@ def forward(ctx, _input, weight, target, bias=None, ignore_index=-100):
...
             grad_logits_chunk = logits_chunk * (
```
Comment (ByronHsu) [inline src/liger_kernel/ops/fused_linear_cross_entropy.py:125]: now we do the loss scaling inside the kernel, the scaling doesn't look right here

**Review by ByronHsu — COMMENTED** [review]
We need to do more testing for flce.

**Comment by FermiQ** [comment]
@CodiumAI-Agent /describe

**Comment by FermiQ** [comment]
@CodiumAI-Agent /improve
